### DISCUSSION TITLE
How to embed Magick.NET to executable?

### DISCUSSION THREAD

**Discussion by TetraTheta** [dlemstra/Magick.NET#1554] [Help] [answered]
I'm trying to create a console app using .NET Framework 4.8, with Visual Studio Community 2022 (17.9.0). OS is Windows 11 x64.  
Currently, I'm only targetting Windows 11 x64, not Windows Server or other OS. But I might want to support Windows x86 later, so I choose AnyCPU variant.

In the app, [Magick.NET-Q8-AnyCPU](https://www.nuget.org/packages/Magick.NET-Q8-AnyCPU) is used to get image's width with this code.
```csharp
public static int GetImageWidth(string path) {
  try {
    MagickImageInfo info = new MagickImageInfo(path);
    if (info.Width <= 0) {
      Console.Error.WriteLine("\x1b[31mERROR\x1b[0m Invalid width of image: '" + path + "'");
      Environment.Exit(1);
      return 0;
    } else {
      return info.Width;
    }
  } catch (Exception ex) {
    Console.Error.WriteLine("\x1b[31mERROR\x1b[0m Failed to get width of image: '" + path + "'");
    Console.Error.WriteLine(ex);
    Environment.Exit(1);
    return 0;
  }
}
```

Since I want to produce a single executable (no DLLs with executable to properly run), I merge all dependencies with [ILMerge](https://www.nuget.org/packages/ILMerge) with [MSBuild.ILMerge.Task](https://www.nuget.org/packages/MSBuild.ILMerge.Task).

ILMerge packs all dependencies that 'Local Copy' is set to 'True' into executable built.  
Magick.NET's 'Local Copy' is also marked as 'True', so ILMerge will merge Magick.NET DLLs too, like it does to other dependencies, and it does. I see increased executable file size from 246KB to 2,500KB because of it.

But while other packed dependencies works as intended, Magick.NET does not.
```plaintext
System.TypeInitializationException: The type initializer for 'NativeMagickSettings' threw an exception. ---> System.DllNotFoundException: Unable to load DLL 'Magick.Native-Q8-x86.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
   at ImageMagick.Environment.NativeMethods.X86.Environment_Initialize()
   at ImageMagick.Environment.NativeEnvironment.Initialize() in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   at ImageMagick.Environment.Initialize() in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   at ImageMagick.MagickSettings.NativeMagickSettings..cctor() in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   --- End of inner exception stack trace ---
   at ImageMagick.MagickSettings.NativeMagickSettings..ctor()
   at ImageMagick.MagickSettings..ctor() in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   at ImageMagick.MagickImage..ctor() in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   at ImageMagick.MagickImageInfo.Read(String fileName, IMagickReadSettings`1 readSettings) in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   at ImageMagick.MagickImageInfo..ctor(String fileName, IMagickReadSettings`1 readSettings) in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   at ImageMagick.MagickImageInfo..ctor(String fileName) in E:\REPO-HDD\cw\Libraries\NaturalSort.cs:line 0
   at cw.Libraries.MiscLibrary.GetImageWidth(String path) in E:\REPO-HDD\cw\Libraries\MiscLibrary.cs:line 46
```

I thought that DLL files are not properly merged, so I added these lines into `.csproj` file.
```xml
<PropertyGroup>
  <MagickCopyNativeWindows>true</MagickCopyNativeWindows>
  <MagickCopyNativeLinux>true</MagickCopyNativeLinux>
  <MagickCopyNativeLinuxMusl>true</MagickCopyNativeLinuxMusl>
  <MagickCopyNativeMacOS>true</MagickCopyNativeMacOS>
</PropertyGroup>
  ```
But this didn't change anything. Executable file size is same and I get same error with before.

I checked if `path` is malformed, but it isn't. I guess something is wrong with ILMerge configuration.

If I remove ILMerge, Magick.NET copies all DLLs into build directory, and works without any problem, but I don't want to put addtional DLL files with my executable, so ILMerge (or [Costura.Fody](https://www.nuget.org/packages/Costura.Fody)) must be used.

This is for MSBuild.ILMerge.Task. It doesn't work.
```json
{
  "General": {
    "InputAssemblies": [
      "CommandLine.dll",
      "Magick.Native-Q8-x64.dll",
      "Magick.Native-Q8-x86.dll",
      "Magick.NET.Core.dll",
      "Magick.NET-Q8-AnyCPU.dll"
    ]
  }
}
```

For Costura.Fody, I tried with `FodyWeaver.xml` but didn't work.
```xml
<Weavers xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="FodyWeavers.xsd">
  <Costura>
    <Unmanaged32Assemblies>
      Magick.Native-Q8-x86
    </Unmanaged32Assemblies>
    <Unmanaged64Assemblies>
      Magick.Native-Q8-x64
    </Unmanaged64Assemblies>
  </Costura>
</Weavers>
```

How to fix this error?

**Comment by dlemstra** [comment]
You will need that extra Native library because this is native code instead. There might be tools for this but I am not aware of any tools that can do this.

**Reply by TetraTheta** [reply]
Strange thing is, I can clearly see `Magick.Native-Q8-x86.dll`, `Magick.Native-Q8-x64.dll` and other DLLs on build directory, if I remove ILMerge dependencies for testing. I also explicitly include them via ILMerge or Fody configuration. (See above for both configuration)

I really don't understand why this happens...

**Reply by dlemstra** [reply, marked as answer]
You cannot merge these `.dll` files with ILMerge because that are native (`C`) instead of `C#`. They will need to be in the same folder as your executable.

**Reply by TetraTheta** [reply]
Ah, I finally get what you mean. Well, if then, I can't merge those DLL files... Thank you for letting me know about that!
